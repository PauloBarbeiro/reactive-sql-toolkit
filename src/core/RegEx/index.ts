/**
 * Module with helper functions to execute Regular Expressions
 */

/**
 * Returns the result of the String.match method against the RegEx:
 * `^(SELECT).+(?<table>${tables.join('|')})`
 *
 * @param query     Query to be tested
 * @param tables    List of database tables
 * @return the result of query.match method
 */
export const readingQueryMatch = (query: string, tables: Array<string>): RegExpMatchArray | null => {
    const readRegEx = new RegExp(`^(SELECT).+(?<table>${tables.join('|')})`)
    return query.match(readRegEx)
}

/**
 * Returns the result of the String.match method against the RegEx:
 * `^(INSERT INTO).+(?<table>${tables.join('|')})`
 *
 * @param query     Query to be tested
 * @param tables    List of database tables
 * @return the result of query.match method
 */
export const writingQueryMatch = (query: string, tables: Array<string>): RegExpMatchArray | null => {
    const readRegEx = new RegExp(`^(INSERT INTO).+(?<table>${tables.join('|')})`)
    return query.match(readRegEx)
}

/**
 * Returns the table on which the query was SELECTing data from.
 *
 * @param query     Query to be tested
 * @param tables    List of database tables
 * @return The table name, or null
 */
export const tableFromReadQuery = (query: string, tables: Array<string>): string | null => {
    const regExRes = readingQueryMatch(query, tables)
    const table = regExRes?.groups?.table
    if(table) {
        return table
    }

    return null
}

/**
 * Returns the table on which the query was INSERTing data to.
 *
 * @param query     Query to be tested
 * @param tables    List of database tables
 * @return The table name, or null
 */
export const tableFromWritingQuery = (query: string, tables: Array<string>): string | null => {
    const regExRes = writingQueryMatch(query, tables)
    const table = regExRes?.groups?.table
    if(table) {
        return table
    }

    return null
}