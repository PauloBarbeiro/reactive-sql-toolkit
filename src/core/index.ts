export { createSQL, destroyDatabase } from './SQL'