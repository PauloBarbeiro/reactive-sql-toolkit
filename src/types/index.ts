import { Database } from "sql.js";

export enum WasmSources {
    prod = 'https://cdnjs.cloudflare.com/ajax/libs/sql.js/1.6.1/sql-wasm.wasm',
    debug = 'https://cdnjs.cloudflare.com/ajax/libs/sql.js/1.6.1/sql-wasm-debug.wasm',
    test = 'node_modules/sql.js/dist/sql-wasm.wasm'
}

/**
 * Defines the Singleton holding the database instance.
 */
export interface DatabaseHolder {
    // The database instance
    instance: Database | null
    // Setter of the instance
    setInstance: (bd: Database) => void
    // Getter of the instance
    getInstance: () => Database | null
    // Instance destructor
    destroy: () => void
}

export type SqlDataType =
    | 'NULL'
    | 'REAL'
    | 'INTEGER'
    | 'INT'
    | 'TEXT'
    | 'BLOB'
    | 'DATE'
    | 'TRUE'
    | 'FALSE'

/**
 * Type that defines a function callback to be used in the Schema values,
 * to generate the INSERT INTO commands. The function will receive the whole
 * value, plus its index in the values array.
 */
export interface ValueFunction {
    asFunc: boolean,
    value: string
}

/**
 * Structure that defines the schema definition of a table.
 * 'fields' will generate the table columns.
 * 'values' will generate INSERT INTO commands.
 */
export type TableDefinitions = {
    fields: Record<string, SqlDataType>
    values?: Array<Record<string, string | number | ValueFunction>>
}

/**
 * Structure that defines the database schema.
 * Each key of the 'Record' refers to a table name, and the TableDefinitions
 * defines the initial construction of the table.
 */
export type Schema = Record<string, TableDefinitions>

export type Function = (args: any) => any
export type Functions = Record<string, Function>

/**
 * Defines the Recorder type.
 * Each string in the Record is associated with a table.
 * Each weakRef item is connected to a 'reactive' function that must be fired
 * after an insert query.
 */
export type Recorder = Record<string, Array<WeakRef<(t:number) => void>>>