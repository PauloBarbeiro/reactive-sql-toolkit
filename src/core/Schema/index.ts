import {Schema, ValueFunction} from "../../types";

/**
 * Create a SQL query from the input schema model.
 *
 * @param schema    Schema for the database
 *
 * @return A tuple with the query string and a list of the tables.
 */
export const createQueryFromSchema = (schema: Schema): [string, Array<string>] => {
    const tables = Object.keys(schema)

    const createTable = "CREATE TABLE"

    let query = ""

    tables.forEach(table => {
        const { fields, values } = schema[table]

        const fieldsPart = Object.keys(fields).reduce((acc, field, idx, keys) => {
            return acc + `${field} ${fields[field]}` + (idx < keys.length - 1 ? ', ' : '')
        }, '')

        const insertPart = values
            ? values.map((data) => {
                const dataPart = Object.keys(data).reduce((acc, key, idx, keys) => {
                    const value = typeof data[key] === "string"
                        ? `'${data[key]}'`
                        : (data[key] as ValueFunction).asFunc
                            ? `${(data[key] as ValueFunction).value}`
                            : `${data[key]}`

                    return acc += (value) + ((idx < keys.length - 1 ? ', ' : ''))
                }, '')

                return `INSERT INTO ${table} VALUES (${dataPart});`
            }).join('')
            : []

        query += `${createTable} ${table} (${fieldsPart});${insertPart}`
    })

    return [query, tables]
}