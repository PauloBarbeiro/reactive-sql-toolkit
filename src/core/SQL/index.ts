import initSqlJs, { Database, BindParams, QueryExecResult } from 'sql.js'

// Types
import {
    DatabaseHolder, Functions,
    Recorder,
    Schema,
} from "../../types";

// Utils
import { createQueryFromSchema } from "../Schema";
import {
    tableFromReadQuery,
    tableFromWritingQuery,
    writingQueryMatch
} from "../RegEx";

/**
 * Object that holds the database instance, and provides basic instance
 * management.
 */
export const database: DatabaseHolder = {
    instance: null,
    setInstance: function(db) {
        this.instance = db
    },
    getInstance: function() {
        return this.instance
    },
    destroy: function () {
        this.instance?.close()
        this.instance = null
    }
}

/**
 * Getter interface to the database.
 */
export const getDatabase = (): Database | null => database.getInstance()

/**
 * Public interface to destroy the database.
 * It executes the db.close() function, and remove it from the singleton holder.
 */
export const destroyDatabase = (): void => {
    database.destroy()
}

/**
 * List of the tables of the database.
 * Used to identify the Read/Write queries.
 */
const tables: Array<string> = []

/**
 * Object holding the weakRefs "listeners".
 */
const GlobalRecorder: Recorder = {}

/**
 * Create SQLite instance using sql.js, from the input 'path' to the wasm file,
 * and schema definition.
 *
 *
 * @param path      Path to the wasm file
 * @param schema    Schema definition to create the database
 * @param functions Functions map to be added to the database
 * @return database or undefined
 */
export const createSQL = async (path: string, schema: Schema, functions?: Functions) => {
    try {
        const buffer = schema.dataBuffer
            ? new Uint8Array(schema.dataBuffer)
            : undefined

        const SQL = await initSqlJs({
            locateFile: () => path
        })

        const db: Database = new SQL.Database(buffer)
        database.setInstance(db)

        if(functions) {
            Object.keys(functions).forEach(name => {
                db.create_function(name, functions[name])
            })
        }

        const [query, tablesList] = createQueryFromSchema(schema)
        tables.push(...tablesList)
        db.exec(query)

        return db
    } catch (e) {
        console.error(e)
    }
    return
}

/**
 * Registers the input 'updateState' function in the 'recorder' object
 * for the related table.
 *
 * - The table's name is selected from the query;
 * - The table, and/or the 'updateState' function, are added to the Recorder
 *
 * @param updateState   Function to be recorder as Actuator
 * @param query         SQL query that was executed
 * @param tables        List of tables of the database
 * @param recorder      Recorder object containing the weakRefs
 * @return void
 */
export const registerQueryListeners = (
    updateState: (time: number) => void,
    query: string,
    tables: Array<string>,
    recorder: Recorder,
):void => {
    const foundTables = tableFromReadQuery(query, tables)

    if(foundTables) {
        foundTables.forEach(table => {
            if(!recorder[table]) {
                recorder[table] = []
            }

            recorder[table].push(new WeakRef<(t: number) => void>(updateState))
        })
    }
}

/**
 * Execute a call to all actuators registered to a certain table.
 *
 * - The table used for the query is selected;
 * - A timestamp is generated;
 * - All actuators, related to the selected table, are executed
 *   asynchronously with the timestamp.
 *
 * @param query     SQL query that was executed
 * @param tables    List of tables of the database
 * @param recorder  Recorder object containing the weakRefs
 * @return void
 */
export const triggerActuators = (
    query: string,
    tables: Array<string>,
    recorder: Recorder,
): void => {
    const table = tableFromWritingQuery(query, tables)

    if(table) {
        const timestamp = Date.now()
        if(!recorder[table]) {
            return
        }

        new Promise((resolve, reject) => {
            try{
                recorder[table].forEach(weakRef => {
                    const fnRef = weakRef.deref()
                    fnRef && fnRef(timestamp)
                })
                resolve(1)
            } catch (err) {
                console.error('UPDATE ERROR: ', err)
                reject()
            }
        })
    }
}

/**
 * Execute the input 'query' with its 'params' in the database instance
 * using the 'exec' interface.
 *
 * @param query     SQL query to be executed
 * @param params    The BindParams to be used in the exec method.
 * @return the database result or undefined.
 */
export const executeQuery = (query: string, params?: BindParams): Array<QueryExecResult> | undefined => {
    let db = getDatabase()

    if(!db) {
        console.error('SQL-lite instance not initiated! \nRun createSQL function to initialize the service.')
        return;
    }

    try {
        return db.exec(query, params)
        // @ts-ignore
    } catch (error: Error) {
        console.error(`SQL-lite Error: ${error.message}`)
        return
    }
}

/**
 * Executes a SQL query for SELECT purpose, and returns its result.
 *
 * - The updateStateFn is register among the listeners;
 * - The Query is executed;
 * - The Actuators are triggered;
 * - the query result is returned.
 *
 * @param updateStateFn     The updater function to register in the GlobalRecorder
 * @param query             SQL query with the INSERT command
 * @param params            BindParams for the query
 * @return The result from 'executeQuery' function (Array<QueryExecResult> | undefined)
 */
export const queryPipeline = (
    updateStateFn: (time: number) => void,
    query: string,
    params?: BindParams,
): Array<QueryExecResult> | undefined => {
    registerQueryListeners(updateStateFn, query, tables, GlobalRecorder)
    const result = executeQuery(query, params)
    triggerActuators(query, tables, GlobalRecorder)

    return result
}

/**
 * Executes a SQL query for INSERT purpose, and returns its result.
 *
 * - The query containing the INSERT command is validated;
 * - The Query is executed;
 * - The Actuators are triggered;
 * - the query result is returned.
 *
 * @param query     SQL query with the INSERT command
 * @param params    BindParams for the query
 * @param recorder  Recorder object (GlobalRecorder as default)
 * @return The result from 'executeQuery' function (Array<QueryExecResult> | undefined)
 */
export const insertQueryPipeline = (
    query: string,
    params?: BindParams,
    recorder: Recorder = GlobalRecorder
): Array<QueryExecResult> | undefined => {
    const regExRes = writingQueryMatch(query, tables)
    if(!regExRes) {
        return
    }

    const result = executeQuery(query, params)
    triggerActuators(query, tables, recorder)
    return result
}

/**
 * Functions to be exposed to the API.
 */
export default {
    createSQL,
    destroyDatabase,
    queryPipeline,
    insertQueryPipeline,
}