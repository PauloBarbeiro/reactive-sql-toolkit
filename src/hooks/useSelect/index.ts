import { useState } from 'react'
import { queryPipeline } from '../../core/SQL'

/**
 * Hook that provide read-only access to the database
 *
 * @param query     Query to be executed
 * @param depths
 */
export function useSelect(query: string, depths?: Array<string | number| boolean | undefined | null>) {
    const [updatedAt, setUpdatedAt] = useState<number>()

    return  queryPipeline(setUpdatedAt, query)
}
