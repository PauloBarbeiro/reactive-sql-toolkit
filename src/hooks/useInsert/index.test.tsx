import React from "react";
import {getByText, render, fireEvent, waitFor} from '@testing-library/react'
import {Schema, WasmSources} from '../../types'
import {createSQL} from '../../core/SQL'
import {useInsert} from "./index";
import {useSelect} from "../useSelect";

const TestComponent = () => {
    const result = useSelect("SELECT age,name FROM test")
    const insertFunction = useInsert()

    const handleAddLennon = () => {
        insertFunction("INSERT INTO test VALUES (3, 22, 'John');")
    }

    const handleAddGeorge = () => {
        insertFunction("INSERT INTO test VALUES (4, 25, 'George');")
    }

    if(!result || result.length === 0) {
        return (
            <p>No data available</p>
        )
    }

    const [data] = result

    return (
        <>
            <table>
                <caption>Table results</caption>
                <thead>
                <tr>
                    {data.columns.map((column, index) => (<th scope="col" key={index}>{column}</th>))}
                </tr>
                </thead>
                <tbody>
                {data.values.map(([age, name], index) => (
                    <tr key={index}>
                        <th scope="row">{name}</th>
                        <td>{age}</td>
                    </tr>
                ))}
                </tbody>
            </table>
            <button onClick={handleAddLennon}>Add Lennon</button>
            <button onClick={handleAddGeorge}>Add Harrison</button>
        </>
    )
}

describe('useInsert', function () {
    beforeAll(async () => {
        const schema: Schema = {
            test: {
                fields: {id: 'INTEGER', age: 'INTEGER', name: 'TEXT'},
                values: [{id: 1, age: 17, name: 'Ringo'}, {id: 2, age: 18, name: 'Paul'}]
            }
        }

        await createSQL(WasmSources.test, schema)
    })

    it('should execute a query and return some data', () => {
        const { container, getByText } = render(<TestComponent />)
        expect(container).toMatchSnapshot()
        expect(getByText('Ringo')).toBeTruthy()
        expect(getByText('Paul')).toBeTruthy()
    })

    it('should react to a change in the queried table (add John Lennon)', async () => {
        const { container, getByText } = render(<TestComponent />)
        expect(container).toMatchSnapshot()

        const addJohnBtn = getByText('Add Lennon')

        fireEvent.click(addJohnBtn)

        await waitFor(() => {
            expect(getByText('John')).toBeTruthy()
        })
    })

    it('should react to a change in the queried table (add George Harrison)', async () => {
        const { container, getByText } = render(<TestComponent />)
        expect(container).toMatchSnapshot()

        const addJohnBtn = getByText('Add Harrison')

        fireEvent.click(addJohnBtn)

        await waitFor(() => {
            expect(getByText('George')).toBeTruthy()
        })
    })
});
