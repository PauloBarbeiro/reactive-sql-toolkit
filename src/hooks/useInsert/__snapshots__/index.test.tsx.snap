// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useInsert should execute a query and return some data 1`] = `
<div>
  <table>
    <caption>
      Table results
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          age
        </th>
        <th
          scope="col"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Ringo
        </th>
        <td>
          17
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Paul
        </th>
        <td>
          18
        </td>
      </tr>
    </tbody>
  </table>
  <button>
    Add Lennon
  </button>
  <button>
    Add Harrison
  </button>
</div>
`;

exports[`useInsert should react to a change in the queried table (add George Harrison) 1`] = `
<div>
  <table>
    <caption>
      Table results
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          age
        </th>
        <th
          scope="col"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Ringo
        </th>
        <td>
          17
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Paul
        </th>
        <td>
          18
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          John
        </th>
        <td>
          22
        </td>
      </tr>
    </tbody>
  </table>
  <button>
    Add Lennon
  </button>
  <button>
    Add Harrison
  </button>
</div>
`;

exports[`useInsert should react to a change in the queried table (add John Lennon) 1`] = `
<div>
  <table>
    <caption>
      Table results
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          age
        </th>
        <th
          scope="col"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Ringo
        </th>
        <td>
          17
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Paul
        </th>
        <td>
          18
        </td>
      </tr>
    </tbody>
  </table>
  <button>
    Add Lennon
  </button>
  <button>
    Add Harrison
  </button>
</div>
`;
